csrc/logical_domain_map.cpp

```diff
@@ -74,32 +74,54 @@ PairwiseLogicalDomainMap::PairwiseLogicalDomainMap(
 
 namespace {
 
-// Returns a producer ID that is indirectly accessed. A bool is also
-// returned indicating there's a corresponding consumer ID. For
-// example, select doesn't have a consumer ID, whereas index_select
-// does.
-std::pair<IterDomain*, bool> getIndexedDomainInfo(
+// Returns producer IDs that don't map identically to consumer. A bool is
+// returned indicating whether corresponding consumer IDs exists. For example,
+// select doesn't have a consumer ID, whereas index_select does.
+std::pair<std::unordered_set<IterDomain*>, bool> getNonMappingDomainInfo(
     const TensorView* producer_tv,
     const TensorView* consumer_tv) {
-  IterDomain* indexed_id = nullptr;
+  std::unordered_set<IterDomain*> non_mapping_ids;
   bool has_consumer_id = false;
   if (auto sop = dynamic_cast<SelectOp*>(consumer_tv->definition())) {
-    indexed_id = sop->getIndexedID();
+    // indexed ID is indirectly accessed
+    non_mapping_ids.insert(sop->getIndexedID());
     has_consumer_id = false;
   } else if (
       auto sop = dynamic_cast<IndexSelectOp*>(consumer_tv->definition())) {
+    // indexed ID is indirectly accessed
     if (producer_tv == sop->lookupTv()) {
-      indexed_id = sop->getIndexedID();
+      non_mapping_ids.insert(sop->getIndexedID());
       has_consumer_id = true;
     }
   } else if (auto gop = dynamic_cast<GatherOp*>(consumer_tv->definition())) {
+    // indexed ID is indirectly accessed
     if (producer_tv == gop->lookupTv()) {
-      indexed_id = gop->getIndexedID();
+      non_mapping_ids.insert(gop->getIndexedID());
+      has_consumer_id = true;
+    }
+  } else if (
+      auto iaop =
+          dynamic_cast<IndexPutAccumulateOp*>(consumer_tv->definition())) {
+    // see [ Note -- IndexPutAccumulateOp semantics ]
+    if (producer_tv == iaop->indexTv()) {
+      // non-broadcast IDs of index tv do not map to output.
+      std::for_each(
+          producer_tv->getLogicalDomain().begin(),
+          producer_tv->getLogicalDomain().end(),
+          [&non_mapping_ids](IterDomain* id) {
+            if (!id->isBroadcast()) {
+              non_mapping_ids.insert(id);
+            }
+          });
+      has_consumer_id = true;
+    } else if (producer_tv == iaop->valueTv()) {
+      // indexing ID of value tv do not map to output.
+      non_mapping_ids.insert(iaop->getIndexingIDOfValue());
```

Comment:
If `seq` is just a size-one dimension (i.e., a broadcast), shouldn't the `seq` dimension of the value tensor be mapped with the consumer?

Reply:
I'm not sure that's the right thing to do.

Semantically, they mean different things. Since the potential mapping output dimension would be `vocab`. I don't think it's right to map the two together.

i.e. if we are going to codegen, we'll want the loop to be resolved to
```
for seq:
  for hidden:
```

wouldn't want to map `seq` to `vocab`, even when `seq` is a broadcast.
I think we do want to map the two `seq` on `index_tv` and `value_tv`, but I'm not sure what is the place to do so.

Reply:
I think I asked this question because when `seq` is size one, that domain of the index tensor is mapped. And I agree that:

> we do want to map the two seq on index_tv and value_tv

So, doesn't it mean the seq dimension of the value tensor should also be mapped?

Reply:
> because when seq is size one, that domain of the index tensor is mapped. And I agree that:

No it's not. From the code comment earlier

```
 // logical ID groups of IndexPutAccumulateOp
  // args:
  //     acc   [ vocab, hidden ]
  //     index [ seq, 1 ]
  //     value [ seq, hidden ]
  // output:
  //     out   [ vocab, hidden ]
```

We don't map `seq` to output at all.

right now we map:
1. everything from `acc` to `out`;
2. `hidden` from `value` to `out`;
3. nothing from `index` to `out`;

specifically, `seq` doesn't show up in `out` at all, so we don't map it. I think the one thing that is missing is that, we *should* map `seq` between `index` and `value`. But I'm not sure where to do that.

Reply:
Hmm, am I reading this part incorrectly?

```
std::for_each(
          producer_tv->getLogicalDomain().begin(),
          producer_tv->getLogicalDomain().end(),
          [&non_mapping_ids](IterDomain* id) {
            if (!id->isBroadcast()) {
              non_mapping_ids.insert(id);
            }
          });
```

If an ID is broadcast, it's not included in `non_mapping_ids`, so I suppose it may be mapped.